src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -51,133 +51,74 @@ CollectionNotifier::get_modification_checker(TransactionChangeInfo const& info,
         };
     }
 
+    // If the table in question has no outgoing links it will be the only entry in `m_related_tables`.
+    // In this case we do not need a `DeepChangeChecker` and check the modifications against the
+    // `ObjectChangeSet` within the `TransactionChangeInfo` for this table directly.
     if (m_related_tables.size() == 1) {
-        auto& object_set = info.tables.find(m_related_tables[0].table_key.value)->second;
+        auto& object_chenge_set = info.tables.find(m_related_tables[0].table_key.value)->second;
         return [&](ObjectChangeSet::ObjectKeyType object_key) {
-            return object_set.modifications_contains(object_key);
+            if (all_callbacks_filtered()) {
+                return object_chenge_set.modifications_contains(object_key, get_filtered_column_keys(true));
+            }
+            else {
+                return object_chenge_set.modifications_contains(object_key, {});
+            }
         };
     }
 
-    return DeepChangeChecker(info, *root_table, m_related_tables);
+    return DeepChangeChecker(info, *root_table, m_related_tables, get_key_path_arrays());
 }
 
-void DeepChangeChecker::find_related_tables(std::vector<RelatedTable>& out, Table const& table)
+std::vector<KeyPathArray> CollectionNotifier::get_key_path_arrays()
 {
-    auto table_key = table.get_key();
-    if (any_of(begin(out), end(out), [=](auto& tbl) {
-            return tbl.table_key == table_key;
-        }))
-        return;
-
-    // We need to add this table to `out` before recurring so that the check
-    // above works, but we can't store a pointer to the thing being populated
-    // because the recursive calls may resize `out`, so instead look it up by
-    // index every time
-    size_t out_index = out.size();
-    out.push_back({table_key, {}});
-
-    for (auto col_key : table.get_column_keys()) {
-        auto type = table.get_column_type(col_key);
-        if (type == type_Link || type == type_LinkList) {
-            out[out_index].links.push_back({col_key.value, type == type_LinkList});
-            find_related_tables(out, *table.get_link_target(col_key));
+    std::vector<KeyPathArray> key_path_arrays = {};
+    for (auto callback : m_callbacks) {
```

Comment:
This copies each of the callbacks while iterating over the vector. The iteration variable for ranged based for should nearly always be `auto&&`.

Reply:
Fixed with 7284b25f0.